### PR TITLE
Depthwise separable convolution

### PR DESCRIPTION
**Pull request by abadams**
This is an interesting app to add to the benchmark suite, because it's a generally useful neural network layer. It's also an interesting case because it's tempting to fuse the depthwise part of the conv into the pointwise part. I spent the day scheduling it, and for me this fusion paid off on CPU but not on GPU.

### CODE REVIEW

File: apps/depthwise_separable_conv/depthwise_separable_conv_generator.cpp [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+            int tile_d = 1;
+            const int vec = natural_vector_size<float>();
+
+            // Figure out how many registers we have in the register
+            // file on this target.
+            int num_regs = 16;
+            if (get_target().has_feature(Target::AVX512_Skylake) ||
+                (get_target().arch == Target::ARM &&
+                 get_target().bits == 64)) {
+                num_regs = 32;
```
Comment (dsharletg) [inline apps/depthwise_separable_conv/depthwise_separable_conv_generator.cpp:205]: Maybe this should be a helper function of Target, similar to natural_vector_size?

File: apps/depthwise_separable_conv/depthwise_separable_conv_generator.cpp [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+            pointwise_filter.dim(0).set_estimate(0, CO);
+            pointwise_filter.dim(1).set_estimate(0, CI * CM);
+
+            bias.dim(0).set_estimate(0, CO);
+
+            output.dim(0).set_estimate(0, CO);
+            output.dim(1).set_estimate(0, W);
+            output.dim(2).set_estimate(0, H);
+            output.dim(3).set_estimate(0, N);
+        } else if (false && get_target().has_gpu_feature()) {
```
[the remark this answers is not in the document]
Comment (abadams) [reply apps/depthwise_separable_conv/depthwise_separable_conv_generator.cpp]: oops, fixed

File: apps/depthwise_separable_conv/depthwise_separable_conv_generator.cpp [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        // Convolve the image depthwise -- for each input channel,
+        // generate channel_multiplier number of intermediate channels using convolution
+        Func depthwise_convolved("depthwise_convolved");
+        Expr pad_width = depthwise_filter.dim(2).extent() / 2;
+        Expr pad_height = depthwise_filter.dim(3).extent() / 2;
+        RDom depthwise_filter_dom(0, depthwise_filter.dim(0).extent(),
+                                  0, depthwise_filter.dim(2).extent(),
+                                  0, depthwise_filter.dim(3).extent());
+        depthwise_convolved(d, x, y, b) +=
+            depthwise_filter(depthwise_filter_dom[0],
```
[the remark this answers is not in the document]
Comment (abadams) [reply apps/depthwise_separable_conv/depthwise_separable_conv_generator.cpp]: done

**Comment by abadams** [comment]
As a side-note, now that I've used it, the version of tile that takes vectors was definitely a good idea.

**Comment by abadams** [comment]
Failures unrelated.

**Pull request merged**